### ISSUE TITLE
Turi create memory error-> MemoryError: std::bad_alloc  in GPU & CPU machine

### ISSUE THREAD

**Issue by dhgokul** [apple/turicreate#615]
I have trained turi create model in CPU machine for 4k dataset and got memory error 

Total dataset size : 103.7 MB, Single image size in KB only .

**Machine configuration :**
![image](https://user-images.githubusercontent.com/5689804/40827195-875f7c62-659a-11e8-9dff-1de2a0a6507b.png)

I have referred memory error already posted in https://github.com/apple/turicreate/issues/267 and updated batch_size parameter as 32 and still got same issue 

**Memory error :**
![image](https://user-images.githubusercontent.com/5689804/40827296-d4fc1764-659a-11e8-99bc-b8f8b9535dc5.png)

**Modified batch size changes :** 
![image](https://user-images.githubusercontent.com/5689804/40827373-0a37ca0e-659b-11e8-91a0-43039ba9df67.png)

**Comment by dhgokul** [comment]
I have run turicreate script in GPU (AWS Deep learning -> g2.2xlarge configuration) and got bad alloc error for same dataset. PFA

![image](https://user-images.githubusercontent.com/5689804/40839091-9df808ee-65be-11e8-8cbb-3f8b3ab11845.png)

Analyzed object_detector code for batch_size, it seems value is updated as 32  

File location : /usr/local/lib/python2.7/dist-packages/turicreate/toolkits/object_detector/object_detector.py
![image](https://user-images.githubusercontent.com/5689804/40839121-bd9df6fe-65be-11e8-8f1a-9100679430cc.png)

Please let me know anything needs to be done

**Comment by dhgokul** [comment]
@gustavla Any update on this issue ?

**Comment by srikris** [comment]
@dhgokul The batch size you modified was for image classifier. You need to do it for OD. Can you try it with

```
model = tc.object_dector.create(dataset, 'annotation', _advanced_parameters = {'batch_size': 16})
```

**Comment by gustavla** [comment]
I don't think this is a batch size issue, since this happens before neural network training begins. I think the issue is that you have large images in the SFrame, and we are not handling that as well as we should. A temporary work-around is to try to resize them before passing them into `create`. The network takes images in the size 416x416, so having your images around that size is ideal. Obviously, we are actively looking for a better solution. Sorry for the inconvenience!

**Comment by gustavla** [comment]
I just saw that you said single images are on KB only, so they are not large after all (at least not in file size). That is a bit more surprising. Let me ask you this then. What is the largest image in terms of width x height area in the dataset? That is, it can be small in file size, but large in dimensions (an example of an image that could fit such description is an image with large flat regions, with lots of potential to compress the image).

**Comment by dhgokul** [comment]
Thanks for your response guys !

@srikris  I checked object_detector code too for batch_size, it seems value is updated as 32
File location : /usr/local/lib/python2.7/dist-packages/turicreate/toolkits/object_detector/object_detector.py

@gustavla  We have four sets of dimension in our dataset - PFA
1280 X 720 (Max dimension)
960 X 720
404 X 720
![dimension](https://user-images.githubusercontent.com/5689804/41115354-de82779c-6aa4-11e8-8a47-bcc4f28d9d0e.png)

**Comment by gustavla** [comment]
@dhgokul Thanks for the info! These are not particularly large images indeed. 

Not sure if this will help, but #384 has some discussion around memory errors and suggests the following:
```python
tc.config.set_runtime_config('TURI_FILEIO_MAXIMUM_CACHE_CAPACITY', 2*1024*1024*1024)
```
Let me know if you find any success adjusting this value. Thanks!

**Comment by srikris** [comment]
@dhgokul @gustavla It might be easier to share the dataset and script so we can debug. Sorry for the trouble!

**Comment by gustavla** [comment]
I think to reproduce this and related memory issues, the key is probably the environment as opposed to the specific dataset. I will set up an environment with less RAM and see if I can repro.

**Comment by dhgokul** [comment]
Thanks @gustavla & @srikris  , will try below configuration update for memory issue and let you know

Meanwhile I like to share list of issues, I had faced during turicreate training process in GPU.

Run turicreate with 859 images and it’s successfully completed and generated coreml model
**Successfully turicreate log:** 
![turicreatelog](https://user-images.githubusercontent.com/5689804/41145577-321ba490-6b1e-11e8-9ed5-41c3ea04abb1.png)

Then I increase dataset (~1900 images) and got different issues mentioned below at each time run turicreate python script
It keeps on loading for long time (~1 to 2 hours )

**Issue 1:** 
![issue1](https://user-images.githubusercontent.com/5689804/41145586-3b453b6c-6b1e-11e8-9a0c-347451916d3d.png)

**Issue 2 :** 
![issue2](https://user-images.githubusercontent.com/5689804/41145595-42c6eb24-6b1e-11e8-8560-12913db66a27.png)

**Issue 3 :** 
![issue3](https://user-images.githubusercontent.com/5689804/41145599-48d3fe1c-6b1e-11e8-8018-e6620c2d84f3.png)

**Comment by dhgokul** [comment]
@gustavla & @srikris  . Posted issue and attached dataset,turicreate python script in apple bug creator. Please refer below link 

https://bugreport.apple.com/web/?problemID=40926208

Summary : 

Area:
CoreML
Summary: Turicreate training is not proceed forward after few iterations(say 91 iteration) @ GPU

Steps to Reproduce: Run turicreate python script in AWS deep learning machine, and it got hanged not proceed forward , But it running in background for an hour

Expected Results: It will create coreML model at the completion of turicreate training process.

Actual Results: Turicreate training is not proceed forward after few iterations(say 91 iteration)

Version/Build: Turi version 4.3.2 (stable version)
cuda 9.0
mxnet 1.1.0

Configuration: AWS deep learning machine (Instance type : g2.2xlarge)

**Comment by dhgokul** [comment]
@gustavla Modified turicreate config for memory changes in GPU and got memory bad exception issue
 tc.config.set_runtime_config('TURI_FILEIO_MAXIMUM_CACHE_CAPACITY', 2*1024*1024*1024)
**Turicreate GPU issue** 
![turicreate_error_gpu](https://user-images.githubusercontent.com/5689804/41164965-0fd84352-6b5a-11e8-9365-87ec204a039b.png)

**Comment by dhgokul** [comment]
Configured turicreate beta version in CPU and run script for same dataset ,got below issue

Attached turicreate training log in CPU - PFA "turicreatebetaversion.txt

[turicreatebetaversion.txt](https://github.com/apple/turicreate/files/2084884/turicreatebetaversion.txt)


Predicting 1951/1951
Traceback (most recent call last):
  File "turi.py", line 81, in <module>
    print('Mean Average Precision(mAP): {:.1%}'.format(scores['mean_average_precision']))
KeyError: 'mean_average_precision'

**Comment by gustavla** [comment]
@dhgokul Really sorry you are having so many issues. I think I spot two distinct issues that I can guess the cause for:

1. It seems to fail shuffling your SFrame, which we know has high memory overhead (#608). In that other issue, did you try turning off shuffling? Did it help?
2. It looks like you might be running out of memory training on the GPU. Do you know how many GB of memory your GPU has? You can check this with `nvidia-smi`. Right now, I believe we require 4 GB, but we do not check this explicitly so that we can handle it more gracefully (we should). Ideally, we should automatically re-configure the batch size to make it fit. For now, you can try things like `create(..., _advanced_parameters={'batch_size': 8})` (the default is 32, try lowering until errors disappear), but we can't guarantee the max iteration selection will be appropriate in this case.

**Comment by dhgokul** [comment]
@gustavla thanks ! I have checked GPU memory using watch nvidia- smi command. It created python process and memory percent increase and decrease for sometime. After that it utilizes CPU process with memory less than 50 %. As of now I can able to create model with less accuracy without any issue by disable shuffle option in create function.

**Comment by dhgokul** [comment]
f we disable shuffle option as false, it will created coreML model, but it is not accurate and detect all the objects when compared to model created with shuffle option as true.

If we enable shuffle option, still we get same memory error mentioned below

[14:02:01] src/operator/nn/./cudnn/./cudnn_algoreg-inl.h:107: Running performance tests to find the best convolution algorithm, this can take a while... (setting env variable MXNET_CUDNN_AUTOTUNE_DEFAULT to 0 to disable)
2018-06-12 14:02:12 Training 1/1000 Loss 4.299
2018-06-12 14:02:22 Training 10/1000 Loss 4.283
2018-06-12 14:02:32 Training 19/1000 Loss 4.101
2018-06-12 14:02:43 Training 28/1000 Loss 3.865
2018-06-12 14:02:53 Training 37/1000 Loss 3.529
2018-06-12 14:03:03 Training 46/1000 Loss 3.121
2018-06-12 14:03:14 Training 54/1000 Loss 3.017
2018-06-12 14:03:25 Training 63/1000 Loss 2.951
2018-06-12 14:03:35 Training 72/1000 Loss 2.947
2018-06-12 14:03:48 Training 81/1000 Loss 2.941
2018-06-12 14:03:58 Training 91/1000 Loss 2.940
Traceback (most recent call last):
File "turi.py", line 73, in
model = tc.object_detector.create(train_data, feature='image', annotations='annotations', max_iterations=1000)
File "/home/ubuntu/venv/local/lib/python2.7/site-packages/turicreate/toolkits/object_detector/object_detector.py", line 336, in create
for batch in loader:
File "/home/ubuntu/venv/local/lib/python2.7/site-packages/turicreate/toolkits/object_detector/_sframe_loader.py", line 123, in next
return self._next()
File "/home/ubuntu/venv/local/lib/python2.7/site-packages/turicreate/toolkits/object_detector/_sframe_loader.py", line 232, in _next
self.sframe = self.sframe.sort(_TMP_COL_RANDOM_ORDER)
File "/home/ubuntu/venv/local/lib/python2.7/site-packages/turicreate/data_structures/sframe.py", line 5421, in sort
return SFrame(_proxy=self.proxy.sort(sort_column_names, sort_column_orders))
File "/home/ubuntu/venv/local/lib/python2.7/site-packages/turicreate/cython/context.py", line 49, in exit
raise exc_type(exc_value)
MemoryError: std::bad_alloc

**Comment by dhgokul** [comment]
@gustavla same memory issue @ CPU with shuffle option = true, 

2018-06-12 20:56:08  Training    1/1000  Loss  4.674
2018-06-12 20:56:19  Training    2/1000  Loss  4.629
2018-06-12 20:56:30  Training    3/1000  Loss  4.595
2018-06-12 20:56:40  Training    4/1000  Loss  4.563
2018-06-12 20:56:51  Training    5/1000  Loss  4.533
2018-06-12 20:57:02  Training    6/1000  Loss  4.496
2018-06-12 20:57:13  Training    7/1000  Loss  4.486
2018-06-12 20:57:23  Training    8/1000  Loss  4.470
2018-06-12 20:57:37  Training    9/1000  Loss  4.440
2018-06-12 20:57:48  Training   10/1000  Loss  4.416
2018-06-12 20:57:59  Training   11/1000  Loss  4.362
2018-06-12 20:58:10  Training   12/1000  Loss  4.328
2018-06-12 20:58:20  Training   13/1000  Loss  4.289
2018-06-12 20:58:31  Training   14/1000  Loss  4.278
2018-06-12 20:58:42  Training   15/1000  Loss  4.281
2018-06-12 20:58:53  Training   16/1000  Loss  4.240
2018-06-12 20:59:06  Training   17/1000  Loss  4.228
2018-06-12 20:59:17  Training   18/1000  Loss  4.189
2018-06-12 20:59:27  Training   19/1000  Loss  4.121
2018-06-12 20:59:38  Training   20/1000  Loss  4.094
2018-06-12 20:59:49  Training   21/1000  Loss  4.072
2018-06-12 21:00:00  Training   22/1000  Loss  4.033
2018-06-12 21:00:11  Training   23/1000  Loss  3.990
2018-06-12 21:00:22  Training   24/1000  Loss  3.966
2018-06-12 21:00:34  Training   25/1000  Loss  3.956
2018-06-12 21:00:47  Training   26/1000  Loss  3.915
2018-06-12 21:00:59  Training   27/1000  Loss  3.897
2018-06-12 21:01:10  Training   28/1000  Loss  3.858
2018-06-12 21:01:21  Training   29/1000  Loss  3.826
2018-06-12 21:01:32  Training   30/1000  Loss  3.761
2018-06-12 21:01:43  Training   31/1000  Loss  3.723
2018-06-12 21:01:54  Training   32/1000  Loss  3.672
2018-06-12 21:02:07  Training   33/1000  Loss  3.626
2018-06-12 21:02:18  Training   34/1000  Loss  3.592
2018-06-12 21:02:29  Training   35/1000  Loss  3.560
2018-06-12 21:02:40  Training   36/1000  Loss  3.531
2018-06-12 21:02:51  Training   37/1000  Loss  3.540
2018-06-12 21:03:02  Training   38/1000  Loss  3.485
2018-06-12 21:03:13  Training   39/1000  Loss  3.481
2018-06-12 21:03:24  Training   40/1000  Loss  3.412
2018-06-12 21:03:36  Training   41/1000  Loss  3.354
2018-06-12 21:03:47  Training   42/1000  Loss  3.318
2018-06-12 21:03:58  Training   43/1000  Loss  3.255
2018-06-12 21:04:09  Training   44/1000  Loss  3.199
2018-06-12 21:04:20  Training   45/1000  Loss  3.134
2018-06-12 21:04:31  Training   46/1000  Loss  3.140
2018-06-12 21:04:42  Training   47/1000  Loss  3.080
2018-06-12 21:04:53  Training   48/1000  Loss  3.095
2018-06-12 21:05:05  Training   49/1000  Loss  3.068
2018-06-12 21:05:16  Training   50/1000  Loss  3.082
2018-06-12 21:05:27  Training   51/1000  Loss  3.058
2018-06-12 21:05:38  Training   52/1000  Loss  3.033
2018-06-12 21:05:49  Training   53/1000  Loss  3.000
2018-06-12 21:06:00  Training   54/1000  Loss  3.030
2018-06-12 21:06:11  Training   55/1000  Loss  3.011
2018-06-12 21:06:22  Training   56/1000  Loss  3.022
2018-06-12 21:06:34  Training   57/1000  Loss  2.988
2018-06-12 21:06:46  Training   58/1000  Loss  2.950
2018-06-12 21:06:57  Training   59/1000  Loss  2.899
2018-06-12 21:07:07  Training   60/1000  Loss  2.880
2018-06-12 21:07:19  Training   61/1000  Loss  2.893
2018-06-12 21:07:31  Training   62/1000  Loss  2.859
2018-06-12 21:07:44  Training   63/1000  Loss  2.822
2018-06-12 21:07:55  Training   64/1000  Loss  2.788
2018-06-12 21:08:09  Training   65/1000  Loss  2.798
2018-06-12 21:08:20  Training   66/1000  Loss  2.826
2018-06-12 21:08:32  Training   67/1000  Loss  2.828
2018-06-12 21:08:42  Training   68/1000  Loss  2.880
2018-06-12 21:08:53  Training   69/1000  Loss  2.904
2018-06-12 21:09:04  Training   70/1000  Loss  2.946
2018-06-12 21:09:15  Training   71/1000  Loss  2.931
2018-06-12 21:09:26  Training   72/1000  Loss  2.884
2018-06-12 21:09:39  Training   73/1000  Loss  2.850
2018-06-12 21:09:51  Training   74/1000  Loss  2.833
2018-06-12 21:10:02  Training   75/1000  Loss  2.838
2018-06-12 21:10:13  Training   76/1000  Loss  2.850
2018-06-12 21:10:24  Training   77/1000  Loss  2.892
2018-06-12 21:10:35  Training   78/1000  Loss  2.910
2018-06-12 21:10:46  Training   79/1000  Loss  2.948
2018-06-12 21:10:57  Training   80/1000  Loss  2.922
2018-06-12 21:11:11  Training   81/1000  Loss  2.921
2018-06-12 21:11:22  Training   82/1000  Loss  2.903
2018-06-12 21:11:33  Training   83/1000  Loss  2.867
2018-06-12 21:11:44  Training   84/1000  Loss  2.923
2018-06-12 21:11:55  Training   85/1000  Loss  2.996
2018-06-12 21:12:06  Training   86/1000  Loss  3.011
2018-06-12 21:12:17  Training   87/1000  Loss  3.012
2018-06-12 21:12:28  Training   88/1000  Loss  3.001
2018-06-12 21:12:40  Training   89/1000  Loss  2.982
2018-06-12 21:12:51  Training   90/1000  Loss  2.970
2018-06-12 21:13:03  Training   91/1000  Loss  2.894
2018-06-12 21:13:14  Training   92/1000  Loss  2.921
Traceback (most recent call last):
  File "turi.py", line 70, in <module>
    model = tc.object_detector.create(train_data, feature='image', annotations='annotations', max_iterations=1000)
  File "/home/aximsoft/venv/local/lib/python2.7/site-packages/turicreate/toolkits/object_detector/object_detector.py", line 336, in create
    for batch in loader:
  File "/home/aximsoft/venv/local/lib/python2.7/site-packages/turicreate/toolkits/object_detector/_sframe_loader.py", line 123, in next
    return self._next()
  File "/home/aximsoft/venv/local/lib/python2.7/site-packages/turicreate/toolkits/object_detector/_sframe_loader.py", line 232, in _next
    self.sframe = self.sframe.sort(_TMP_COL_RANDOM_ORDER)
  File "/home/aximsoft/venv/local/lib/python2.7/site-packages/turicreate/data_structures/sframe.py", line 5421, in sort
    return SFrame(_proxy=self.__proxy__.sort(sort_column_names, sort_column_orders))
  File "/home/aximsoft/venv/local/lib/python2.7/site-packages/turicreate/cython/context.py", line 49, in __exit__
    raise exc_type(exc_value)
MemoryError: std::bad_alloc

**Comment by dhgokul** [comment]
@gustavla  Based on my observation turicreate trained  dataset up to 1094 images and can create coreML mode, if we train more than that, it return memory error

Do you have any update on this issue ?

**Comment by gustavla** [comment]
@dhgokul We are looking into this, but it doesn't have a quick fix so it might take some time until we roll out a fix. Apologies for this!

**Comment by afranklin** [comment]
@gustavla Do we have a repro for this?

**Comment by gustavla** [comment]
@afranklin Not yet. I have tried with an ubuntu VM where I could control and reduce RAM. I still could not repro this even with only 2 GB RAM and a large SFrame.

**Comment by dabarparihar** [comment]
I faced the similar issue while training for ~7k images. After analysing I found that there was a mismatch of the class labels provided during training with the labels in data set that was present in .csv file which I used for creating sframe. After fixing the label in dataset I recreated sframe and it worked for me.

**Comment by gustavla** [comment]
@dabarparihar That's interesting - thanks for reporting this! I'm not sure why mismatching the labels would lead to a memory failure like this. So, the only difference between the two datasets was that one had effectively random labels and the other one had correct ones? Were there any other changes?

**Comment by shantanuchhabra** [comment]
@dhgokul Just as an update, we still haven't been able to repro this specific issue. We did repro #608 and fixed it with #904 which is now merged into master. 

This issue looks very similar to #608 from the logs and we think it may be fixed by #904 . Could you try building Turi Create from master to verify that? We also plan to release Turi Create 5.0beta3 soon so if you want to wait for the Turi Create wheel, that works too!

**Comment by shantanuchhabra** [comment]
Hi @dhgokul , 

We just released Turi Create 5.0 beta 3, you can find the wheels on the [release page](https://github.com/apple/turicreate/releases). We think this should be fixed now but please let us know if you can still repro the issue and we'll reopen it!

**Comment by dhgokul** [comment]
Thanks ! Will check and let you know
